main.go

```diff
@@ -616,6 +308,499 @@ func main() {
+		})
+		// Logs
+		if cfg.logs.enabled {
+			r.Group(func(r chi.Router) {
+				r.Use(authentication.WithTenant(t.Name, t.ID))
+				r.Use(authN)
+				r.Use(authentication.WithTenantHeader(cfg.logs.tenantHeader, t.ID))
+				r.Mount("/api/logs/v1/"+t.Name, stripTenantPrefix("/api/logs/v1", logsHandler(cfg, tCfg, authZ)))
+			})
+		}
+
+		level.Info(tCfg.logger).Log("msg", "tenant registration is successful :", t.Name)
+
+		break
+	}
+}
+
+func registerTenantRetryMetric(tCfg *tenantsConfig) (*prometheus.CounterVec, error) {
+	retryCounterMetric := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "tenant_onboarding_attempts_total",
+			Help: "Number of tenant onboarding attempts.",
+		},
+		[]string{"tenant"},
+	)
+
+	err := tCfg.reg.Register(retryCounterMetric)
+	if err != nil {
+		level.Info(tCfg.logger).Log("msg", "duplicate registration of metric", "error", err)
+		return retryCounterMetric, err
+	}
+
+	return retryCounterMetric, nil
+}
+
+// Handler for Metrics legacy API.
+func metricsLegacyHandler(cfg *config, tCfg *tenantsConfig, authZ rbac.Authorizer) http.Handler {
+	return metricslegacy.NewHandler(
+		cfg.metrics.readEndpoint,
+		metricslegacy.Logger(tCfg.logger),
+		metricslegacy.Registry(tCfg.reg),
+		metricslegacy.HandlerInstrumenter(tCfg.ins),
+		metricslegacy.SpanRoutePrefix("/api/v1/{tenant}"),
+		metricslegacy.ReadMiddleware(authorization.WithAuthorizers(authZ, rbac.Read, "metrics")),
+		metricslegacy.ReadMiddleware(authorization.WithEnforceTenantLabel(cfg.metrics.tenantLabel)),
+		metricslegacy.UIMiddleware(authorization.WithAuthorizers(authZ, rbac.Read, "metrics")))
+}
+
+// Handler for Metrics V1 API.
+func metricsHandler(cfg *config, tCfg *tenantsConfig, authZ rbac.Authorizer) http.Handler {
+	return metricsv1.NewHandler(
+		cfg.metrics.readEndpoint,
+		cfg.metrics.writeEndpoint,
+		metricsv1.Logger(tCfg.logger),
+		metricsv1.Registry(tCfg.reg),
+		metricsv1.HandlerInstrumenter(tCfg.ins),
+		metricsv1.SpanRoutePrefix("/api/metrics/v1/{tenant}"),
+		metricsv1.ReadMiddleware(authorization.WithAuthorizers(authZ, rbac.Read, "metrics")),
+		metricsv1.ReadMiddleware(authorization.WithEnforceTenantLabel(cfg.metrics.tenantLabel)),
+		metricsv1.WriteMiddleware(authorization.WithAuthorizers(authZ, rbac.Write, "metrics")),
+	)
+}
+
+// Handler for Logs V1 API.
+func logsHandler(cfg *config, tCfg *tenantsConfig, authZ rbac.Authorizer) http.Handler {
+	return logsv1.NewHandler(
+		cfg.logs.readEndpoint,
+		cfg.logs.tailEndpoint,
+		cfg.logs.writeEndpoint,
+		logsv1.Logger(tCfg.logger),
+		logsv1.Registry(tCfg.reg),
+		logsv1.HandlerInstrumenter(tCfg.ins),
+		logsv1.SpanRoutePrefix("/api/logs/v1/{tenant}"),
+		logsv1.ReadMiddleware(authorization.WithAuthorizers(authZ, rbac.Read, "logs")),
+		logsv1.WriteMiddleware(authorization.WithAuthorizers(authZ, rbac.Write, "logs")))
+}
+
+// Calculate exponential backoff wait time for retrying tenant onboarding.
```

Comment:
```suggestion
// expBackOffWaitTime calculate exponential backoff wait time,
```
I don't think it does anything around tenancy, it's pretty generic, right?